### PR TITLE
MQTT Client:  Added support for receiving empty payload (IDFGH-946)

### PR DESCRIPTION
**Pull request by gregjesl**
Closes https://github.com/espressif/esp-idf/issues/3088

### CODE REVIEW

File: mqtt_client.c [hunk 1]
```diff
@@ -530,6 +536,7 @@ static esp_err_t esp_mqtt_dispatch_event(esp_mqtt_client_handle_t client)
 
 static void deliver_publish(esp_mqtt_client_handle_t client, uint8_t *message, int length)
 {
+    mqtt_header_state_t mqtt_header_state = MQTT_HEADER_STATE_INCOMPLETE;
     const char *mqtt_topic = NULL, *mqtt_data = NULL;
     uint32_t mqtt_topic_length, mqtt_data_length;
     uint32_t mqtt_len = 0, mqtt_offset = 0, total_mqtt_len = 0;
```
Comment (david-cermak) [inline mqtt_client.c]: It looks like `total_mqtt_len` is not needed anymore

File: mqtt_client.c [hunk 2]
```diff
@@ -540,19 +547,21 @@ static void deliver_publish(esp_mqtt_client_handle_t client, uint8_t *message, i
...
+                /* mqtt header is not complete, continue reading */
                 memmove(client->mqtt_state.in_buffer, message, length);
                 buffer_offset = length;
                 message = client->mqtt_state.in_buffer;
                 max_to_read = client->mqtt_state.in_buffer_length - length;
                 mqtt_len = 0;
             } else {
+                /* mqtt header was fully read */
+                mqtt_header_state = MQTT_HEADER_STATE_COMPLETE;
                 total_mqtt_len = client->mqtt_state.message_length - client->mqtt_state.message_length_read + mqtt_data_length;
```
Comment (david-cermak) [inline mqtt_client.c]: also this line could be removed

**Comment by gregjesl** [comment]
@david-cermak I've made the changes, please let me know if there is anything else.

**Comment by david-cermak** [comment]
@gregjesl LGTM, thanks. 

However Travis has failed, will have to fix it and then ask for a rebase from your side.

**Comment by david-cermak** [comment]
@gregjesl Could you please rebase now? Build would probably still fail as it's also validated against IDF master (which might not currently build on some platforms: linux with newer python packages), but there's not an issue in this repo and changes can be reviewed internally in the meantime.

**Comment by gregjesl** [comment]
@david-cermak You'd like me to rebase the `idf` branch onto this pull request, correct?

**Comment by david-cermak** [comment]
@gregjesl Yes, please rebase your branch (in this pull request) on top of `idf`. That's then easier for us to test and review, plus don't have to modify your commits

**Comment by gregjesl** [comment]
@david-cermak I've rebased the branch.

**Comment by david-cermak** [comment]
@gregjesl There seems to be an issue with receiving empty messages under some conditions (e.g. single message with empty payload in a buffer).

Sorry I haven't noticed during code review and haven't had time to test it before. Conditional `NULL == (mqtt_data = mqtt_get_publish_data(message, &mqtt_data_length)` seems to be a culprit as this could be null for empty messages.

**Comment by gregjesl** [comment]
From the MQTT 3.1.1 spec:
> The length of the payload can be calculated by subtracting the length of the variable header from the Remaining Length field that is in the Fixed Header.

The length of the variable header is simply the topic length + 4.  I can't find the part of the code that returns the Remaining Length field.

**Comment by david-cermak** [comment]
Generally, I agree that would be the correct solution. There's no API which directly returns "Remaining Length" (also note the variable header length also depends on QOS).

With the given codebase however, I might prefer to modify `mqtt_get_publish_data` (and maybe the topic function) to indicate if a starting payload ptr could be resolved. Something like:
```
bool mqtt_get_publish_data(uint8_t* buffer, uint32_t* length, char** mqtt_data);

```

**Comment by gregjesl** [comment]
@david-cermak The following function should return `true` if `buffer` contains a complete message: 
```cpp
bool mqtt_message_complete(uint8_t* buffer, uint32_t blength)
{
    int i;
    int remainlen = 0;

    // Read out the Remaining Length (which does not include the length of the fixed header)
    for (i = 1; i < 5; ++i)
    {
        if(i >= blength)
            return false;
        remainlen += (buffer[i] & 0x7f) << (7 * (i - 1));
        if ((buffer[i] & 0x80) == 0)
        {
            ++i;
            break;
        }
    }
    
    // The buffer length must be at least the Remaining Length 
    // plus the length of the fixed header (i in this case)
    return blength >= remainlen + i;
}
```

The test
``` cpp
if(NULL == (mqtt_topic = mqtt_get_publish_topic(message, &mqtt_topic_length)) || 
    NULL == (mqtt_data = mqtt_get_publish_data(message, &mqtt_data_length)))
```
can be replaced with
``` cpp
if(!mqtt_message_complete(message, length))
```
The buffer contains a complete message without a payload if
```cpp
mqtt_message_complete(message, length) == true
```
and
```cpp
NULL == (mqtt_data = mqtt_get_publish_data(message, &mqtt_data_length))
```

Would these changes work for you?

**Comment by gregjesl** [comment]
@david-cermak Disregard my last comment, it will fail for large messages.  

This function should return `true` if the header is completely loaded in the buffer:
```cpp
bool mqtt_header_complete(uint8_t* buffer, uint32_t blength)
{
    int i;
    int topiclen;

    for (i = 1; i < 5; ++i)
    {
        if(i >= blength)
            return false;
        if ((buffer[i] & 0x80) == 0)
        {
            ++i;
            break;
        }
    }
    // i is the length of the fixed header

    if (i + 2 >= blength)
        return false;
    topiclen = buffer[i++] << 8;
    topiclen |= buffer[i++];

    if (i + topiclen > blength)
        return false;

    i += topiclen;

    if (mqtt_get_qos(buffer) > 0)
    {
        i += 2;
    }
    // i is now the length of the fixed + variable header
    return blength >= i;
}
```

**Comment by gregjesl** [comment]
@david-cermak I've made the udpates.

**Comment by david-cermak** [comment]
@gregjesl Thanks!
This works fine and doesn't break existing tests. Have moved to internal repo for a review internally.

Meantime, could you please correct:
* if - else formatting per https://docs.espressif.com/projects/esp-idf/en/latest/contribute/style-guide.html#braces
* line 571 -- missing semicolon;

**Comment by gregjesl** [comment]
@david-cermak I've made the formatting changes to `mqtt_client.c` and fixed the semicolon issue.  I did not adjust the formatting in `mqtt_msg.c` as it seems to match the formatting in that specific document.  

Is there a way for me to test my changes before committing them besides altering the git submodule in the IDF? I didn't realize until today that the examples in the examples folder use the MQTT library in my IDF and not the ones in the local repository.

**Comment by david-cermak** [comment]
@gregjesl Thanks, yes I meant only the related changes to `mqtt_client.c`.

The easiest way to test it is to modify the submodule in IDF. You can then run the python tests in example folder. (there are other ways to include an external component, such as `EXTRA_COMPONENT_DIR` -- you can check `.travis` file in this repo to see how it is built, but I do not recommend)

**Comment by david-cermak** [comment]
Just two minor comments arose from the review
* use of that macro
* remove the "closing idf issue" from the commit message (I might have suggested otherwise in the beginning, but agree that the issue is to be closed upon submodule update in esp-idf not esp-mqtt, sorry for that)

@gregjesl I would prefer if you updated your changes, as it could be merged directly without me amending the commit.

File: lib/mqtt_msg.c [hunk 3]
```diff
@@ -172,6 +173,38 @@ uint32_t mqtt_get_total_length(uint8_t* buffer, uint16_t length)
     return totlen;
 }
 
+bool mqtt_header_complete(uint8_t* buffer, uint16_t buffer_length)
+{
+    uint16_t i;
+    uint16_t topiclen;
+
+    for (i = 1; i < 5; ++i)
```
Comment (david-cermak) [inline lib/mqtt_msg.c]: Macro `MQTT_MAX_FIXED_HEADER_SIZE` could be used here?

**Comment by gregjesl** [comment]
@david-cermak That sounds good; I'll make the changes later today.

**Comment by gregjesl** [comment]
@david-cermak I've made the changes.

**Comment by david-cermak** [comment]
@gregjesl Many thanks for your contribution!

**Pull request merged**